processor/metricstarttimeprocessor/internal/datapointstorage/timeseries_map.go

```diff
@@ -0,0 +1,103 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package datapointstorage // import "github.com/open-telemetry/opentelemetry-collector-contrib/processor/metricstarttimeprocessor/internal/datapointstorage"
+
+import (
+	"sync"
+
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/pmetric"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/pdatautil"
+)
+
+// TimeseriesInfo contains the information necessary to adjust from the initial point and to detect resets.
+type TimeseriesInfo struct {
+	Mark bool
+
+	Number               pmetric.NumberDataPoint
+	Histogram            pmetric.HistogramDataPoint
+	ExponentialHistogram pmetric.ExponentialHistogramDataPoint
+	Summary              pmetric.SummaryDataPoint
+}
+
+type TimeseriesKey struct {
+	Name           string
+	Attributes     [16]byte
+	AggTemporality pmetric.AggregationTemporality
+}
+
+// TimeseriesMap maps from a timeseries instance (metric * label values) to the timeseries info for
+// the instance.
+type TimeseriesMap struct {
+	sync.RWMutex
+	// The mutex is used to protect access to the member fields. It is acquired for the entirety of
+	// AdjustMetricSlice() and also acquired by gc().
+
+	Mark   bool
+	TsiMap map[TimeseriesKey]*TimeseriesInfo
+}
+
+// Get the TimeseriesInfo for the timeseries associated with the metric and label values.
+func (tsm *TimeseriesMap) Get(metric pmetric.Metric, kv pcommon.Map) (*TimeseriesInfo, bool) {
+	// This should only be invoked be functions called (directly or indirectly) by AdjustMetricSlice().
+	// The lock protecting tsm.tsiMap is acquired there.
+	name := metric.Name()
+	key := TimeseriesKey{
+		Name:       name,
+		Attributes: getAttributesSignature(kv),
+	}
+	switch metric.Type() {
+	case pmetric.MetricTypeHistogram:
+		// There are 2 types of Histograms whose aggregation temporality needs distinguishing:
+		// * CumulativeHistogram
+		// * GaugeHistogram
+		key.AggTemporality = metric.Histogram().AggregationTemporality()
+	case pmetric.MetricTypeExponentialHistogram:
+		// There are 2 types of ExponentialHistograms whose aggregation temporality needs distinguishing:
+		// * CumulativeHistogram
+		// * GaugeHistogram
+		key.AggTemporality = metric.ExponentialHistogram().AggregationTemporality()
+	}
+
+	tsm.Mark = true
+	tsi, ok := tsm.TsiMap[key]
+	if !ok {
+		tsi = &TimeseriesInfo{}
+		tsm.TsiMap[key] = tsi
+	}
+	tsi.Mark = true
+	return tsi, ok
+}
+
+// Create a unique string signature for attributes values sorted by attribute keys.
+func getAttributesSignature(m pcommon.Map) [16]byte {
```

Comment:
The 16 length byte array is referenced multiple times here. Maybe a named type alias over `[16]byte` would be helpful for keeping those in sync and conveying the meaning more clearly.